### PR TITLE
`Development`: Move docker image pulls for programming exercises from startup to first build for local continuous integration

### PR DESCRIPTION
**Pull request by laurenzfb**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The pulling of docker images should not happen during the server startup as it delays the whole process. Docker images should be pulled when they are actually needed.

### Description
This PR essentially moves some image pull logic from the Local CI configuration (which is initialized on startup) to the build job management service.

Before a build job is added to the build job queue, Artemis checks whether the image already exists. If not, it is pulled. Thus, the image is pulled the first time a programming exercise using that template is created.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Make sure the image `ls1tum/artemis-maven-template:java17-18` is removed from your local machine
2. Start the LocalVC & LocalCI setup on your local machine
3. Check whether the image is pulled during server startup (it shouldn't be)
4. Create a programming exercise
5. Check whether is pulled during exercise creation

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by pal03377 — CHANGES_REQUESTED** [review]
Very cool! I tested it locally, but I got an error when I tried to open start a programming submission with the online editor open: 

https://github.com/ls1intum/Artemis/assets/9006596/95695d75-e5ce-43a6-9c71-cc191c71a9c7

Probably the online editor indirectly tries to fetch from the local VC system, which is not available yet.

Another concern: What happens if the image is still pulling and I try to start another build while that is happening? Will it pull twice because the image is still not there?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java [hunk 1]
```diff
@@ -53,6 +66,29 @@ public void createBuildPlanForExercise(ProgrammingExercise programmingExercise,
...
+            }
+            catch (InterruptedException ie) {
+                throw new LocalCIException("Interrupted while pulling docker image " + dockerImage, ie);
+            }
+        }
+        catch (BadRequestException e) {
+            throw new LocalCIException("Error while inspecting docker image " + dockerImage, e);
+        }
+
+        // Trigger build for the given participation.
```
Comment (pal03377) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java:90]: Can you reuse the code from `LocalCIBuildJobManagementService` here (or vice versa)? This is a bit duplicated here.

**Review by pal03377 — COMMENTED** [review]
It now worked locally 🎉

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java (hunk 1 above)
Comment (laurenzfb) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java]: Moved the method to a new class.

**Review by pal03377 — APPROVED** [review]
Very nice! 🙌 

Tested locally (also by sending many submissions to test the Docker pull lock) and reviewed the code.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIDockerService.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import de.tum.in.www1.artemis.exception.LocalCIException;
+
+/**
+ * Service for Docker related operations in local CI
+ */
+
+@Service
+@Profile("localci")
+
+public class LocalCIDockerService {
```
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIDockerService.java:24]: Why did you add these extra empty lines here? In Artemis we normally don't add them.
```suggestion
 */
@Service
@Profile("localci")
public class LocalCIDockerService {
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIDockerService.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    /**
+     * Pulls a docker image if it is not already present on the system
+     * Uses a lock to prevent multiple threads from pulling the same image
+     *
+     * @param imageName the name of the docker image
+     */
+
+    public void pullDockerImage(String imageName) {
```
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIDockerService.java:43]: Same here with an extra space.
```suggestion
     */
    public void pullDockerImage(String imageName) {
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIBuildJobManagementService.java [hunk 4]
```diff
@@ -49,19 +49,26 @@ public class LocalCIBuildJobManagementService {
...
+    private final LocalCIDockerService localCIDockerService;
+
     @Value("${artemis.continuous-integration.timeout-seconds:120}")
     private int timeoutSeconds;
 
     @Value("${artemis.continuous-integration.asynchronous:true}")
     private boolean runBuildJobsAsynchronously;
 
+    @Value("${artemis.continuous-integration.build.images.java.default}")
+    String dockerImage;
```
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIBuildJobManagementService.java:61]: You should probably make this private as well.
```suggestion
    private String dockerImage;
```

**Review by Hialus — CHANGES_REQUESTED** [review]
Code looks mostly good to me. I just found 3 small styling issues.

**Comment by laurenzfb** [comment]
Fixed the addressed issues

**Review by pal03377 — DISMISSED** [review]

**Review by Strohgelaender — DISMISSED** [review]
code lgtm

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIDockerService.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import de.tum.in.www1.artemis.exception.LocalCIException;
+
+/**
+ * Service for Docker related operations in local CI
+ */
+@Service
+@Profile("localci")
+
+public class LocalCIDockerService {
+
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIDockerService.java:24]: superfluous line

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIBuildJobManagementService.java [hunk 6]
```diff
@@ -49,19 +49,26 @@ public class LocalCIBuildJobManagementService {
...
 
+    private final LocalCIDockerService localCIDockerService;
+
     @Value("${artemis.continuous-integration.timeout-seconds:120}")
     private int timeoutSeconds;
 
     @Value("${artemis.continuous-integration.asynchronous:true}")
     private boolean runBuildJobsAsynchronously;
 
+    @Value("${artemis.continuous-integration.build.images.java.default}")
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIBuildJobManagementService.java:60]: This could be final if injected in the constructor

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java [hunk 7]
```diff
@@ -40,10 +41,17 @@ public class LocalCIService extends AbstractContinuousIntegrationService {
 
     private final LocalCITriggerService localCITriggerService;
 
+    private final LocalCIDockerService localCIDockerService;
+
+    @Value("${artemis.continuous-integration.build.images.java.default}")
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java:46]: could be final as well

**Review by tobias-lippert — DISMISSED** [review]
Code LGTM. I just have some optional suggestions.

**Review by tobias-lippert — APPROVED** [review]

**Review by pal03377 — APPROVED** [review]

**Review by lennart-keller — APPROVED** [review]
Tested during testing session. Works well

**Review by milljoniaer — APPROVED** [review]
tested in testing session and worked well
